daemon/k8s_watcher.go

```diff
@@ -1632,3 +1675,56 @@ func (d *Daemon) deleteK8sPodV1(pod *v1.Pod) {
 	}
 
 }
+
+func (d *Daemon) addK8sNodeV1(k8sNode *v1.Node) {
+	logger := log.WithFields(logrus.Fields{
+		"K8sNodeName": k8sNode.ObjectMeta.Name,
+	})
+
+	n := k8s.ParseNode(k8sNode)
+	var hostIP net.IP
+	if n.Name != node.GetName() {
+		hostIP = n.GetNodeIP(false)
+	}
+	ip4 := hostIP.To4()
+	if ip4 == nil {
+		logger.Debugf("Skipping ipcache update, HostIP is not an IPv4 address %s", hostIP)
+		return
+	}
+
+	ciliumIPStr := k8sNode.GetAnnotations()[annotation.CiliumHostIP]
+	ciliumIP := net.ParseIP(ciliumIPStr)
+	if ciliumIP == nil {
+		logger.Errorf("no/invalid Cilium-Host IP: %s", ciliumIPStr)
+		return
+	}
+
+	updated := ipcache.IPIdentityCache.Upsert(ciliumIPStr, ipcache.Identity{
+		ID:     identity.ReservedIdentityCluster,
+		Source: ipcache.FromKubernetes,
+	})
+	if !updated {
+		logger.Errorf("ipcache entry owned by kvstore or agent")
+		return
+	}
+
+	id := identity.IPIdentityPair{
+		IP:     ciliumIP,
+		HostIP: hostIP,
+		ID:     identity.ReservedIdentityCluster,
+	}
+
+	for _, listener := range d.ipcacheListeners {
+		listener.OnIPIdentityCacheChange(ipcache.Upsert, nil, id)
+	}
+}
+
+func (d *Daemon) updateK8sNodeV1(_, k8sNode *v1.Node) {
+	// TODO rely on the kvstore integration?
+	return
+}
+
+func (d *Daemon) deleteK8sNodeV1(k8sNode *v1.Node) {
+	// TODO rely on the kvstore integration?
```

Comment:
This needs to do the same as for pod delete events:

```
        // a small race condition exists here as deletion could occur in
        // parallel based on another event but it doesn't matter as the
        // identity is going away
        id, exists := ipcache.IPIdentityCache.LookupByIP(pod.Status.PodIP)
        if !exists {
                return true, fmt.Errorf("identity for IP does not exist in case")
        }

        if id.Source != ipcache.FromKubernetes {
                return true, fmt.Errorf("ipcache entry not owned by kubernetes source")
        }

        ipcache.IPIdentityCache.Delete(pod.Status.PodIP)
```

Reply:
fixed